obspy/io/reftek/core.py

```diff
@@ -0,0 +1,341 @@
+            st = Stream()
+            for channel_number in np.unique(data['channel_number']):
+                inds = data['channel_number'] == channel_number
+                # channel number of EH/ET packets also equals zero (one of the
+                # three unused bytes in the extended header of EH/ET packets)
+                inds &= data['packet_type'] == b"DT"
+                packets = data[inds]
+
+                # split into contiguous blocks, i.e. find gaps. packet sequence
+                # was sorted already..
+                endtimes = (packets[:-1]["time"] +
+                            packets[:-1]["number_of_samples"] * delta)
+                # check if next starttime matches seamless to last chunk
+                # 1e-3 seconds == 1 millisecond is the smallest time difference
+                # reftek130 format can represent, so anything larger or equal
+                # means a gap/overlap.
+                # for now be conservative and check even more rigorous against
+                # 1e-4 to be on the safe side, but in the gapless data example
+                # the differences are always 0 or -2e-7 (for POSIX timestamps
+                # of order 1e9) which seems like a floating point accuracy
+                # issue for np.float64.
+                gaps = np.abs(packets[1:]["time"] - endtimes) > 1e-4
+                if np.any(gaps):
+                    gap_split_indices = np.nonzero(gaps)[0] + 1
+                    contiguous = np.array_split(packets, gap_split_indices)
+                else:
+                    contiguous = [packets]
+
+                for packets_ in contiguous:
+                    starttime = packets_[0]['time']
+
+                    # Unfortunately the whole data cannot be unpacked with one
+                    # call to libmseed as some payloads do not take the full
+                    # 960 bytes. They are thus padded which would results in
+                    # padded pieces directly in a large array and libmseed
+                    # (understandably) does not support that.
+                    #
+                    # Thus we resort to *tada* pointer arithmetics in Python
+                    # ;-) This is quite a bit faster then correctly casting to
+                    # an integer pointer so its worth it.
+                    #
+                    # Also avoid a data copy.
+                    #
+                    # Writing this directly in C would be about 3 times as fast
+                    # so it might be worth it.
+                    npts = packets_["number_of_samples"].sum()
+                    if headonly:
+                        sample_data = np.array([], dtype=np.int32)
+                    else:
+                        sample_data = np.empty(npts, dtype=np.int32)
+                        pos = 0
+                        s = packets_[0]["payload"][40:].ctypes.data
+                        if len(packets_) > 1:
+                            offset = (
```

Comment:
I looked into this a bit more. To answer your question: Yes, the offset is constant assuming all packet payloads are a contiguous chunk in memory, since all packets have same length (1024 bytes) and the sample data in DT packets starts at position 40 into the payload (i.e. at byte 64 -- 24 bytes header plus the empty 40 bytes padding in the payload).

But.. what I currently don't get is why the payloads of the packets are always contiguous in memory. When I print the numpy ndarray flags **it tells me that it's both non-contiguous for C and F** but on the other hand the offset is always the same..:
```python
+ 305                         print(packets_["payload"].flags)
+ 306                         for i in range(1, len(packets_)):
>>307                             print(packets_[i]["payload"].ctypes.data - packets_[i-1]["payload"].ctypes.data)
  308                         sample_data = np.empty(npts, dtype=np.int32)
  309                         pos = 0
  310                         s = packets_[0]["payload"][40:].ctypes.data
  311                         if len(packets_) > 1:
  312                             offset = (
  313                                 packets_[1]["payload"][40:].ctypes.data - s)
  314                         else:
  315                             offset = 0
  316                         for p in packets_:
  317                             _npts = p["number_of_samples"]
  318                             clibmseed.msr_decode_steim1(
  319                                 s, 960, _npts, sample_data[pos:], _npts, None,
  320                                 1) 
  321                             pos += _npts
  322                             s += offset
```

```
  C_CONTIGUOUS : False
  F_CONTIGUOUS : False
  OWNDATA : False
  WRITEABLE : True
  ALIGNED : True
  UPDATEIFCOPY : False
1031
1031
1031
1031
```

I'm considering to make this foolproof by a patch.. 
```diff
diff --git a/obspy/io/reftek/core.py b/obspy/io/reftek/core.py
index db37db1..967cdbd 100644
--- a/obspy/io/reftek/core.py
+++ b/obspy/io/reftek/core.py
@@ -304,19 +304,14 @@ class Reftek130(object):
                     else:
                         sample_data = np.empty(npts, dtype=np.int32)
                         pos = 0
-                        s = packets_[0]["payload"][40:].ctypes.data
-                        if len(packets_) > 1:
-                            offset = (
-                                packets_[1]["payload"][40:].ctypes.data - s)
-                        else:
-                            offset = 0
+                        # Sample data starts at byte 40 in the DT packet's
+                        # payload.
                         for p in packets_:
                             _npts = p["number_of_samples"]
                             clibmseed.msr_decode_steim1(
-                                s, 960, _npts, sample_data[pos:], _npts, None,
-                                1)
+                                p["payload"][40:].ctypes.data, 960, _npts,
+                                sample_data[pos:], _npts, None, 1)
                             pos += _npts
-                            s += offset
 
                     tr = Trace(data=sample_data, header=header.copy())
                     if headonly:
```
to make it
```python
  305                         sample_data = np.empty(npts, dtype=np.int32)
  306                         pos = 0
~ 307                         # Sample data starts at byte 40 in the DT packet's
~_308                         # payload.
  309                         for p in packets_:
  310                             _npts = p["number_of_samples"]
  311                             clibmseed.msr_decode_steim1(
~ 312                                 p["payload"][40:].ctypes.data, 960, _npts,
~ 313                                 sample_data[pos:], _npts, None, 1)
_ 314                             pos += _npts
```

but that makes the benchmark (1 file with 1 hour 3channel 200Hz data, 2MB filesize) drop from 80 to 95 msec.. :-/